src/conmon.c

```diff
@@ -534,31 +534,69 @@ static int write_oom_files()
 	return oom_fd >= 0 ? 0 : -1;
 }
 
-static gboolean oom_cb_cgroup_v1(int fd, GIOCondition condition, G_GNUC_UNUSED gpointer user_data)
+/* user_data is expected to be the container's pid, used to verify the cgroup hasn't been cleaned up */
+static gboolean oom_cb_cgroup_v1(int fd, GIOCondition condition, gpointer user_data)
 {
-	uint64_t oom_event;
+	int *pid = (int *)user_data;
+	uint64_t event_count;
 	ssize_t num_read = 0;
+	_cleanup_free_ char *memory_cgroup_path = NULL;
+	_cleanup_free_ char *cgroup_event_control_path = NULL;
 
-	if ((condition & G_IO_IN) != 0) {
-		num_read = read(fd, &oom_event, sizeof(uint64_t));
-		if (num_read < 0) {
-			nwarn("Failed to read oom event from eventfd");
-			return G_SOURCE_CONTINUE;
-		}
+	if ((condition & G_IO_IN) == 0) {
+		/* End of input */
+		close(fd);
+		oom_event_fd = -1;
+		return G_SOURCE_REMOVE;
+	}
+
+	num_read = read(fd, &event_count, sizeof(uint64_t));
+	if (num_read < 0) {
+		nwarn("Failed to read oom event from eventfd");
+		return G_SOURCE_CONTINUE;
+	}
+
+	if (num_read == 0) {
+		close(fd);
+		oom_event_fd = -1;
+		return G_SOURCE_REMOVE;
+	}
+
+	if (num_read != sizeof(uint64_t)) {
+		nwarn("Failed to read full oom event from eventfd");
+		return G_SOURCE_CONTINUE;
+	}
 
-		if (num_read > 0) {
-			if (num_read != sizeof(uint64_t))
-				nwarn("Failed to read full oom event from eventfd");
-			write_oom_files();
+	ndebugf("Memory cgroup event count: %ld", (long)event_count);
+	if (event_count == 0) {
+		nwarn("Unexpected event count (zero)");
+		return G_SOURCE_CONTINUE;
+	}
+	/* attempt to read the container's cgroup path.
+	 * if we can't, the cgroup has probably been cleaned up.
+	 * In all likelihood, this means we recieved an event on the eventfd
+	 * because the memory.oom_control file was removed, not because of an OOM
+	 */
+	memory_cgroup_path = process_cgroup_subsystem_path(*pid, false, "memory");
+	/* if event_count == 1, we know the only event triggered was a cgroup removal
+	 * if it was greater than 1, we know the cgroup oomed, and then was cleaned up.
```

Comment:
can't there be a race if the cgroup is removed between the read of `event_count` and the time `process_cgroup_subsystem_path` fails?

Reply:
hm, you're right. I changed it so now we get the cgroup.memory_control path directly, so there's less chance of a race ( i don't know if there's anything we can do to entirely mitigate the race, we have to read the fd and check the file and they can't be atomic)

Reply:
yeah this is tricky :/ even reading again from the eventfd won't be enough